### PR TITLE
Add servicebus

### PR DESCRIPTION
**Pull request by helayoty**
Add Service Bus Azure module to terratest

### CODE REVIEW

**Comment by yorinasub17** [comment]
Thanks for your contribution! Can you also add some tests for the functions, either unit or integration?

Note that while we want to add more support for Azure in terratest, we need some help to maintain the code, and we aren't planning on merging Azure contributions until we can figure out a way to run builds on Azure. Please take a look at https://github.com/gruntwork-io/terratest/issues/89#issuecomment-582494195 to learn more.

**Comment by brikis98** [comment]
Please see https://github.com/gruntwork-io/terratest/issues/89 for the status on Terratest with Azure. In particular, we're [starting some work](https://github.com/gruntwork-io/terratest/issues/89#issuecomment-650052374) around it, and when that work is done, we'll be able to come back to this PR!

**Comment by helayoty** [comment]
@HadwaAbdelhalem The PR is ready for review

**Comment by helayoty** [comment]
https://github.com/helayoty/terratest/runs/1780323871?check_suite_focus=true

**Comment by helayoty** [comment]
Issue #771

File: modules/azure/servicebus.go
Comment (HadwaAbdelhalem) [inline modules/azure/servicebus.go:163]: use require.NoError(t, err)

File: modules/azure/servicebus.go
Comment (HadwaAbdelhalem) [inline modules/azure/servicebus.go:192]: use require.NoError(t, err)

File: modules/azure/servicebus.go
Comment (HadwaAbdelhalem) [inline modules/azure/servicebus.go:221]: use require.NoError(t, err)

File: modules/azure/servicebus.go
Comment (HadwaAbdelhalem) [inline modules/azure/servicebus.go:256]: use require.NoError(t, err)

File: modules/azure/servicebus.go
Comment (HadwaAbdelhalem) [inline modules/azure/servicebus.go:291]: use require.NoError(t, err)

File: modules/azure/servicebus.go
Comment (HadwaAbdelhalem) [inline modules/azure/servicebus.go:325]: use require.NoError(t, err)

File: modules/azure/servicebus.go
Comment (HadwaAbdelhalem) [inline modules/azure/servicebus.go:359]: use require.NoError(t, err)

File: modules/azure/servicebus.go
Comment (HadwaAbdelhalem) [inline modules/azure/servicebus.go:394]: use require.NoError(t, err)

File: examples/azure/terraform-azure-servicebus-example/main.tf
Comment (HadwaAbdelhalem) [inline examples/azure/terraform-azure-servicebus-example/main.tf:27]: typo Define

File: examples/azure/terraform-azure-servicebus-example/outputs.tf
Comment (HadwaAbdelhalem) [inline examples/azure/terraform-azure-servicebus-example/outputs.tf:6]: move resource_group to the top of the file to match the rest of the samples

File: modules/azure/servicebus.go
Comment (HadwaAbdelhalem) [inline modules/azure/servicebus.go:322]: general feedback on all like functions, use the same description for functionE

File: test/azure/terraform_azure_servicebus_example_test.go
Comment (HadwaAbdelhalem) [inline test/azure/terraform_azure_servicebus_example_test.go:45]: use ListTopicSubscriptionsName instead since it will require no error, consider naming subscriptionNamesFromAzure to actualsubscriptionNames

File: test/azure/terraform_azure_servicebus_example_test.go
Comment (HadwaAbdelhalem) [inline test/azure/terraform_azure_servicebus_example_test.go:65]: can you use the parseListOfMaps at modules\terraform\output.go instead

**Review by HadwaAbdelhalem — CHANGES_REQUESTED** [review]

File: test/azure/terraform_azure_servicebus_example_test.go
Comment (helayoty) [reply test/azure/terraform_azure_servicebus_example_test.go:65]: This function is private

**Comment by helayoty** [comment]
New Run from CI/CD --> https://github.com/helayoty/terratest/runs/1786762894?check_suite_focus=true

**Review by HadwaAbdelhalem — CHANGES_REQUESTED** [review]
Mostly LGTM, minor last fix to get approved

**Comment by helayoty** [comment]
@HadwaAbdelhalem the latest CI/CD result --> https://github.com/helayoty/terratest/runs/1835742059?check_suite_focus=true

**Review by HadwaAbdelhalem — DISMISSED** [review]
LGTM, @yorinasub17  the PR is ready for you review

File: modules/azure/servicebus.go
Comment (yorinasub17) [inline modules/azure/servicebus.go:179]: NIT: this code block is repeated between `ListServiceBusNamespaceNamesE` and `ListServiceBusNamespaceNamesByResourceGroupE`. Can you create an internal helper that plucks the namespace names given a slice of namespaces?

This comment applies also to `ID` attribute flavor.

File: modules/azure/servicebus.go
Comment (yorinasub17) [inline modules/azure/servicebus.go:45]: Can you switch these slice pointer return values to simple slices (applies to all functions in this file)? Note that slices are already pointers so you can still return `nil`.

In general, we try to avoid usage of pointers, especially pointers to slices as it leads to increased complexity unless it is necessary. For example, in this case you will have to do nil checks whenever you access these slices as `nil` is no longer equivalent to `[]` and the dereference in the `range` calls can lead to panics. Additionally, slice pointers are most useful when you want to pass by reference, and the use case for that is usually limited to function args.

File: modules/azure/servicebus.go
Comment (yorinasub17) [inline modules/azure/servicebus.go:87]: Use [safePtrToString](https://github.com/gruntwork-io/terratest/blob/master/modules/azure/common.go#L56) when dereferencing string ptr attributes.

**Review by yorinasub17 — COMMENTED** [review]
Mostly LGTM! Found a few minor issues that need to be fixed, but once those are addressed I think we can merge this in!

File: modules/azure/servicebus.go
Comment (jtyr) [inline modules/azure/servicebus.go:90]: This should be removed as there is no new `err` definition in the `for` loop.

File: modules/azure/servicebus.go
Comment (jtyr) [inline modules/azure/servicebus.go:118]: This should be removed as there is no new `err` definition in the `for` loop.

File: modules/azure/servicebus.go
Comment (jtyr) [inline modules/azure/servicebus.go:178]: This should be removed as there is no new `err` definition in the `for` loop.

File: modules/azure/servicebus.go
Comment (jtyr) [inline modules/azure/servicebus.go:205]: This should be removed as there is no new `err` definition in the `for` loop.

File: modules/azure/servicebus.go
Comment (jtyr) [inline modules/azure/servicebus.go:305]: This should be removed as there is no new `err` definition in the `for` loop.

**Review by jtyr — CHANGES_REQUESTED** [review]

**Comment by HadwaAbdelhalem** [comment]
Hi @yorinasub17  I updated the PR to address your comments, please review when you get a chace. Here is a link to the [CI ](https://github.com/helayoty/terratest/runs/3592557851?check_suite_focus=true). the failure is related to the ACI test and I will fix in a separate PR.

File: modules/azure/servicebus.go
Comment (yorinasub17) [reply modules/azure/servicebus.go:93]: @HadwaAbdelhalem looks like this comment wasn't addressed. I'm assuming that's because there is some guarantee in the SDK that ensures this is set. Is that correct assumption?

File: modules/azure/servicebus.go
Comment (yorinasub17) [inline modules/azure/servicebus.go:296]: NIT: This should really be `err := iteratorSubscription.Next()`, or even better is:

```
if err := iteratorSubscription.Next(); err != nil {
    return nil, err
}
```

so that it defines a new err var scoped to the forloop/if block. This is a general go pattern to avoid mistakes arising from overriding variables unintentionally.

This applies to the other sections as well.

**Review by yorinasub17 — DISMISSED** [review]
LGTM! Had 2 comments, but they are very minor issues so no need to block the PR.

Let me kick off a sanity check build to make sure it passes our precommit checks and if it does, we can merge this in!

**Comment by yorinasub17** [comment]
Ah looks like the precommit hook failed on running `gofmt`. Here is the diff:

```diff
diff --git a/modules/azure/servicebus_test.go b/modules/azure/servicebus_test.go
index 25c3aad4..f96077dd 100644
--- a/modules/azure/servicebus_test.go
+++ b/modules/azure/servicebus_test.go
@@ -18,7 +18,7 @@ func TestListServiceBusNamespaceNamesE(t *testing.T) {
 
 	subscriptionID := ""
 
-	_, err :=  ListServiceBusNamespaceNamesE(subscriptionID)
+	_, err := ListServiceBusNamespaceNamesE(subscriptionID)
 	require.Error(t, err)
 }
 
@@ -28,7 +28,7 @@ func TestListServiceBusNamespaceIDsByResourceGroupE(t *testing.T) {
 	subscriptionID := ""
 	resourceGroup := ""
 
-	_, err :=  ListServiceBusNamespaceIDsByResourceGroupE(subscriptionID, resourceGroup)
+	_, err := ListServiceBusNamespaceIDsByResourceGroupE(subscriptionID, resourceGroup)
 	require.Error(t, err)
 }
 
@@ -39,7 +39,7 @@ func TestListNamespaceAuthRulesE(t *testing.T) {
 	namespace := ""
 	resourceGroup := ""
 
-	_, err :=  ListNamespaceAuthRulesE(subscriptionID, namespace, resourceGroup)
+	_, err := ListNamespaceAuthRulesE(subscriptionID, namespace, resourceGroup)
 	require.Error(t, err)
 }
 
@@ -50,7 +50,7 @@ func TestListNamespaceTopicsE(t *testing.T) {
 	namespace := ""
 	resourceGroup := ""
 
-	_, err :=  ListNamespaceTopicsE(subscriptionID, namespace, resourceGroup)
+	_, err := ListNamespaceTopicsE(subscriptionID, namespace, resourceGroup)
 	require.Error(t, err)
 }
 
@@ -62,7 +62,7 @@ func TestListTopicAuthRulesE(t *testing.T) {
 	resourceGroup := ""
 	topicName := ""
 
-	_, err :=  ListTopicAuthRulesE(subscriptionID, namespace, resourceGroup, topicName)
+	_, err := ListTopicAuthRulesE(subscriptionID, namespace, resourceGroup, topicName)
 	require.Error(t, err)
 }
 
@@ -76,4 +76,4 @@ func TestListTopicSubscriptionsNameE(t *testing.T) {
 
 	_, err := ListTopicSubscriptionsNameE(subscriptionID, namespace, resourceGroup, topicName)
 	require.Error(t, err)
-}
\ No newline at end of file
+}
```

File: modules/azure/servicebus.go
Comment (HadwaAbdelhalem) [reply modules/azure/servicebus.go:93]: @yorinasub17 it is, the SDK is calling the Azure Rest API underneath for all resources operations, and the validation of the the name as a required parameter is done on the rest API [Here](https://docs.microsoft.com/en-us/rest/api/servicebus/stable/namespaces/create-or-update)

**Review by yorinasub17 — COMMENTED** [review]
Updates LGTM! I'll kick off another sanity check build.

**Review by yorinasub17 — APPROVED** [review]
Passed the precommit hooks! @HadwaAbdelhalem Let me know if this is ready to merge (in particular, if the Azure build passed for you).

**Comment by HadwaAbdelhalem** [comment]
@yorinasub17 it is ready for merge, the [CI ](https://github.com/helayoty/terratest/runs/3704189734?check_suite_focus=true) has a failed test that is unrelated to this PR. it is the TestTerraformAzureACIExample  is failing duo to a change in the MS container image link. I created an [Issue  ](https://github.com/gruntwork-io/terratest/issues/997) to track the ACI failure and will submit a fix for later today

**Comment by yorinasub17** [comment]
Thanks! Will merge this in and release!

**Pull request merged**
